### PR TITLE
Allow indentation of here-string closing characters.

### LINKED ISSUE #2337: here string syntax
Issue
----------------
it would be nice it here strings could close without necessarily needing to be at the start of a line.  Specifically it would be nice if I could tab out the closing side of a here string to to match the opening indentation.

ex:
```powershell
  @"
    Some wacky
    Multiline
    String
  "@
```
instead of this:
```powershell
  @"
    Some wacky
    Multiline
    String
"@
```

Environment data
----------------

```powershell
Name                           Value
----                           -----
PSVersion                      6.0.0-alpha
PSEdition                      Core
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   3.0.0.0
GitCommitId                    v6.0.0-alpha.9
CLRVersion
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1

```

### PR DESCRIPTION
**Pull request by sethvs**
## PR Summary

Fix #2337 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed - Issue link: PowerShell/PowerShell-Docs#2471
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by iSazonov** [comment]
@sethvs Please add tests.

**Comment by sethvs** [comment]
Added tests and docs.

File: test/powershell/Language/Parser/Parser.Tests.ps1 [hunk 1]
```diff
@@ -784,6 +784,11 @@ foo``u{2195}abc
         { ExecuteCommand "`$herestr=@`"`n'`"'`n`"@" } | Should -Not -Throw
     }
 
+    It "A here string closing characters can be indented with spaces" {
+        { ExecuteCommand "@`"`n   here`n   string`n   `"@"} | Should -Not -Throw
```
Comment (iSazonov) [inline test/powershell/Language/Parser/Parser.Tests.ps1]: Please remove the line - next test do this.

Comment (sethvs) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: Agree. Removed.

**Review by iSazonov — APPROVED** [review]

**Comment by lzybkr** [comment]
As mentioned in the issue, this **is** a potential breaking change, so the committee must review.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this and the benefit doesn't justify the breaking change as the PR is currently defined so we cannot accept this as-is.  @sethvs if you want to pursue this, you would need to consider other needs of Here Strings which may require a new token so that this doesn't break existing scripts.

**Comment by joeyaiello** [comment]
To add a couple other points: 

* this doesn't buy you much from a readability perspective when you can't indent the lines between the opening and closing operators, e.g.:

```powershell
function foo {
        $a = @"
Some text
that can't
be indented.
        "@
}
```

* it may be a reasonable user expectation that if they had a trailing amount of whitespace, that it'd be included in the here-string, e.g.

```powershell
function foo {
        $a = @"
Some text
that can't
be indented.
        "@ # why wouldn't the whitespace on this line be included?
}
```

* given that we don't get additional indentation in the content of the here-string, "good enough" readability is achievable with this syntax:

```powershell
function foo {
        $a = 
# inline python code cares about whitespace
@"
if 1 == 1:
    print("yes")
else:
    print("no")
"@
}
```

**Comment by lzybkr** [comment]
My original design years ago (prototyped but never shipped) would strip as much indentation on each line as we found before the terminator - this made here strings quite readable indented, but it is potentially confusing.

**Comment by rkeithhill** [comment]
@lzybkr That's kind of how I would expect any such change to behave. That said, I'm not sure it is worth it at this point.

**Comment by iSazonov** [comment]
@sethvs Thanks for your contribution! 

I close the PR based on PowerShell Committee conclusion. We should continue the discussion in related Issue.

**Comment by SteveL-MSFT** [comment]
@lzybkr we discussed that option (to use the terminator as indicator of stripping leading whitespace), but the main discussion was it would be confusing to the end user.  Perhaps a new token would resolve that.  Personally, I would like to do something like this:

```powershell
  $content = @"
    This is
    my text.
    "@
```

**Comment by iSazonov** [comment]
We could use quotes in last line to set size of the desired indentation:
```c#
$content = @"
        This is
        my text.
"""""@
```

**Pull request closed without merging**